cmd/clusterctl/client/repository/goproxy.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type goproxyClient struct {
+	scheme string
+	host   string
+}
+
+func newGoproxyClient(scheme, host string) *goproxyClient {
+	return &goproxyClient{
+		scheme: scheme,
+		host:   host,
+	}
+}
+
+func (g *goproxyClient) getVersions(owner, repository string) ([]string, error) {
+	// build the naive path
+	var gomodulePath string
+	switch owner {
+	case "kubernetes-sigs":
+		gomodulePath = path.Join("sigs.k8s.io", repository)
+	case "kubernetes":
+		gomodulePath = path.Join("k8s.io", repository)
+	default:
+		gomodulePath = path.Join("github.com", owner, repository)
+	}
+
+	rawURL := url.URL{
+		Scheme: g.scheme,
+		Host:   g.host,
+		Path:   path.Join(gomodulePath, "@v", "/list"),
+	}
+
+	var rawResponse []byte
+	var retryError error
+	_ = wait.PollImmediate(retryableOperationInterval, retryableOperationTimeout, func() (bool, error) {
+		req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, rawURL.String(), http.NoBody)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get versions for go module %q via goproxy %q: failed to create request", gomodulePath, g.host)
```

Comment:
```suggestion
			return false, errors.Wrapf(err, "failed to create request", gomodulePath, g.host)
```
How about this? Becasue the erorr is already wrapped in context in https://github.com/chrischdi/cluster-api/blob/76c3459dcc5b81c3bb3e04cf8170e743e4727ea2/cmd/clusterctl/client/repository/repository_github.go#L113